src/inventory/transaction/EnchantTransaction.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\inventory\transaction;
+
+use pocketmine\block\inventory\EnchantInventory;
+use pocketmine\inventory\transaction\action\InventoryAction;
+use pocketmine\item\Item;
+use pocketmine\item\ItemTypeIds;
+use pocketmine\player\Player;
+use function count;
+
+class EnchantTransaction extends InventoryTransaction{
+
+	private int $optionId;
+	private ?Item $inputItem = null;
+	private ?Item $outputItem = null;
+	private int $lapisCost = 0;
+
+	public function __construct(Player $source, int $optionId){
+		$this->optionId = $optionId;
+		parent::__construct($source);
+	}
+
+	public function validate() : void{
```

Comment:
`getOutput()` is not called in EnchantTransaction, but in ItemStackRequestExecutor. If the optionId is incorrect, `getOutput()` will return null, and EnchantTransaction already has a check if the `outputItem` is null

Reply:
It just doesn't make sense to check the optionId directly in the EnchantTransaction

Reply:
Enchantment level is based on optionId.

Reply:
Yes, enchantment level is optionId + 1, but what kind of validation do you want? In ItemStackRequestExecutor we get the optionId (that's `$recipeId`), and through `EnchantInventory::getOutput(int $optionId)` we get the output item for that optionId. It doesn't make sense for us to somehow validate the optionId in the EnchantTransaction, since EnchantInventory itself checks the validity of optionId, and if optionId is not correct, then `getOutputItem(int $optionId)` returns null, which is what we check in EnchantTransaction. The `getEnchantmentLevel()` in EnchantTransaction is only for convinience, and it returns just `$this->optionId + 1`